### PR TITLE
Added a new Definition to upload firmware images greater than 2GB in size 

### PR DESCRIPTION
**Pull request by aapatwa**
Added a new definition to read the file in chunks, to create chunks URI and to upload firmware

### CODE REVIEW

File: ucsmsdk/ucsgenutils.py [hunk 1]
```diff
@@ -226,6 +226,76 @@ def upload_file(driver, uri, file_dir, file_name, progress=Progress()):
...
+    Returns:
+        None
+
+    Example:
+        driver = UcsDriver()\n
+        upload_firmware(driver=UcsDriver(), uri="http://fileurl", file_dir='/home/user/backup', file_name='my_config_backup.xml')
+    """
+
+    content_path = os.path.join(file_dir, file_name)
+    content_size = os.stat(content_path).st_size
```
Comment (vvb) [inline ucsmsdk/ucsgenutils.py:270]:
```suggestion
    content_size = os.path.getsize(content_path)
```

File: ucsmsdk/ucsgenutils.py [hunk 2]
```diff
@@ -226,6 +226,76 @@ def upload_file(driver, uri, file_dir, file_name, progress=Progress()):
...
+    index = uri.rfind("/")
+    if flag == 0 :
+        res = uri[ : index ] + "/" + str(counter) + uri[index : ]
+        return res
+    else :
+        res = uri[ : index ] + "/merge-" + str(counter) + uri[index : ]
+        return res
+
+
+def upload_firmware(driver, uri, file_dir, file_name, progress=Progress()):
```
Comment (vvb) [inline ucsmsdk/ucsgenutils.py:251]:
```suggestion
def upload_firmware(driver, uri, file_dir, file_name, progress=Progress(), chunk_size=1024*1024):
```

File: ucsmsdk/ucsgenutils.py [hunk 3]
```diff
@@ -226,6 +226,76 @@ def upload_file(driver, uri, file_dir, file_name, progress=Progress()):
...
+    Example:
+        driver = UcsDriver()\n
+        upload_firmware(driver=UcsDriver(), uri="http://fileurl", file_dir='/home/user/backup', file_name='my_config_backup.xml')
+    """
+
+    content_path = os.path.join(file_dir, file_name)
+    content_size = os.stat(content_path).st_size
+
+    f = open(content_path,'rb')
+    CHUNK_SIZE = 10*1024*1024
```
Comment (vvb) [inline ucsmsdk/ucsgenutils.py:273]: Make Chunk_size an argument to this function which has a default value, that can be over-ridden if required.

File: ucsmsdk/ucsgenutils.py (hunk 1 above)
Comment (vvb) [inline ucsmsdk/ucsgenutils.py:270]: Check if the file exists. if not, return an error up ahead itself. Currently, you will see an exception if the file does not exist.

File: ucsmsdk/ucsgenutils.py [hunk 4]
```diff
@@ -226,6 +226,76 @@ def upload_file(driver, uri, file_dir, file_name, progress=Progress()):
...
+    counter = 0
+    flag = 0
+
+    for chunk in read_in_chunks(f,CHUNK_SIZE):
+        if len(chunk) > 0:
+            try:
+                counter += 1
+                uri1 = generate_uri_for_chunks(uri,counter,flag)
```
Comment (vvb) [inline ucsmsdk/ucsgenutils.py:281]: url1 needs a better name

File: ucsmsdk/ucsgenutils.py [hunk 5]
```diff
@@ -226,6 +226,76 @@ def upload_file(driver, uri, file_dir, file_name, progress=Progress()):
...
+                response = driver.post(uri1, data = chunk)
```
Comment (vvb) [inline ucsmsdk/ucsgenutils.py:282]: Do we need to specify any headers here? Content-Type?

File: ucsmsdk/ucsgenutils.py (hunk 4 above)
Comment (vvb) [inline ucsmsdk/ucsgenutils.py:281]: conventions - All `,` need a following space.

**Review by vvb — CHANGES_REQUESTED** [review]

File: ucsmsdk/ucsgenutils.py (hunk 5 above)
Comment (aapatwa) [reply ucsmsdk/ucsgenutils.py:282]: We should not because driver.post() will take care of these headers.

File: ucsmsdk/ucsgenutils.py (hunk 3 above)
Comment (aapatwa) [reply ucsmsdk/ucsgenutils.py:273]: We can change the value based on supported limit , but if we make it as argument then end user can override it with unsupported chunk size which again can cause the same issue that we are trying to fix here

File: ucsmsdk/ucsgenutils.py (hunk 1 above)
Comment (aapatwa) [reply ucsmsdk/ucsgenutils.py]: Yes, I will add this validation here

File: ucsmsdk/ucsgenutils.py (hunk 4 above)
Comment (aapatwa) [reply ucsmsdk/ucsgenutils.py:281]: Sure , replacing uri1 with chunk_uri 
and uri2 with merge_uri

Comment (aapatwa) [reply ucsmsdk/ucsgenutils.py:281]: Sure, added the following space

Comment (vvb) [reply ucsmsdk/ucsgenutils.py:281]: Did you forget to push the commit that does this?

Comment (vvb) [reply ucsmsdk/ucsgenutils.py:281]: I don't see the changes in the code.

Comment (aapatwa) [reply ucsmsdk/ucsgenutils.py]: As per your last email, you said you will review it again.
I was waiting for the same.
I have committed the code now.

**Review by vvb — APPROVED** [review]

File: ucsmsdk/utils/ucsguilaunch.py [hunk 6]
```diff
@@ -56,7 +56,7 @@ def ucs_gui_launch(handle, need_url=False):
 
     jnlp_file = None
     try:
-        ucsm_gui_url = "%s/ucsm/ucsm.jnlp" % handle.uri
```
Comment (vvb) [inline ucsmsdk/utils/ucsguilaunch.py:59]: How is this change related to this PR or firmware upload in general?

Comment (aapatwa) [reply ucsmsdk/utils/ucsguilaunch.py:59]: This is an another issue -
https://github.com/CiscoUcs/ucsmsdk/issues/241
https://jira-eng-sjc11.cisco.com/jira/browse/UCSES-27 (Please refer attachment)

Does it required to create a new Fork and raise PR ?

Comment (vvb) [reply ucsmsdk/utils/ucsguilaunch.py:59]: You can create new branches in your fork for each bug fix and submit individual PRs from each of your bug fix branches. Do not mix multiple unrelated changes in a single PR.

Comment (aapatwa) [reply ucsmsdk/utils/ucsguilaunch.py:59]: Sure,
